### DISCUSSION TITLE
Batch dimensions in linear layers

### DISCUSSION THREAD

**Discussion by mikevitz** [pyg-team/pytorch_geometric#7782] [Q&A]
Hello experts,
I have a class:

```
class GIN(torch.nn.Module):
    def __init__(self, node_features, hidden_channels, output_dim):
        super(GIN, self).__init__()
        self.conv1 = GINConv(
            Sequential(Linear(node_features, hidden_channels),
                       BatchNorm1d(hidden_channels), ReLU(),
                       Linear(hidden_channels, hidden_channels), ReLU()))
        self.conv2 = GINConv(
            Sequential(Linear(hidden_channels, hidden_channels), BatchNorm1d(hidden_channels), ReLU(),
                       Linear(hidden_channels, hidden_channels), ReLU()))
        self.conv3 = GINConv(
            Sequential(Linear(hidden_channels, hidden_channels), BatchNorm1d(hidden_channels), ReLU(),
                       Linear(hidden_channels, hidden_channels), ReLU()))
        self.lin1 = Linear(hidden_channels*3, hidden_channels*3)
        self.lin2 = Linear(hidden_channels*3, output_dim)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        batch = data.batch
        # Node embeddings
        h1 = self.conv1(x, edge_index)
        h2 = self.conv2(h1, edge_index)
        h3 = self.conv3(h2, edge_index)

        # Graph-level readout
        h1 = global_add_pool(h1, batch)
        h2 = global_add_pool(h2, batch)
        h3 = global_add_pool(h3, batch)

        # Concatenate graph embeddings
        h = torch.cat((h1, h2, h3), dim=1)

        # Classifier
        h = self.lin1(h)
        h = h.relu()
        h = F.dropout(h, p=0.5, training=self.training)

        #h = global_mean_pool(h,data.batch)

        h = self.lin2(x)

        return h
```
My dataset is the TUD protein dataset.  For example it looks like:
Number of graphs: 1113
Number of nodes: 47
Number of features: 3
Number of classes: 2

But number of nodes is variable, as each graph is is a different protein.

When I run my training, I get the error:

RuntimeError: mat1 and mat2 shapes cannot be multiplied (2184x3 and 96x2)

How can I resolve this?  I guess the batch dimension is not matching the linear layer input, but the details escape me.

Thank you very much!

**Comment by wsad1** [comment]
Are you getting this error in the line `h = self.lin1(h)` ? can you print out the dimension of the tensors before the line in which you get this error. 
Further when you initialize `self.conv1` you seem to be using `dim_h` ? is `dim_h` same as `hidden_channels`?

**Reply by mikevitz** [reply]
My apologies, I had edited the class before posting.  Yes there is logic setting dim_h = hidden_channels.  (I will just edit the post to remove dim_h and replace with hidden_channels).

I get the error at evaluation of self.lin2   The first linear output layer self.lin1 works fine, and so do all the GINConv layers.  

I am traveling but will post the full error this evening.  Thank you very much!

**Reply by wsad1** [reply]
you need to do `self.lin2(h)` instead of `self.lin2(x)`.

**Reply by mikevitz** [reply]
Thank you!
